controllers/hyperparametertuningjob/hyperparametertuningjob_controller_test.go

```diff
@@ -18,1680 +18,538 @@ package hyperparametertuningjob
-
-		// Verify that the SageMaker request was made
-		Expect(receivedRequests.Len()).To(Equal(1))
-	})
-
-	Context("when the spec is updated", func() {
+		var expectedStatus sagemaker.HyperParameterTuningJobStatus
 
 		BeforeEach(func() {
-
-			// Get job so we can modify it.
-			err = k8sClient.Get(context.Background(), types.NamespacedName{
-				Namespace: job.ObjectMeta.Namespace,
-				Name:      job.ObjectMeta.Name,
-			}, job)
-			Expect(err).ToNot(HaveOccurred())
-
-			job.Spec.HyperParameterTuningJobConfig.ResourceLimits.MaxNumberOfTrainingJobs = ToInt64Ptr(*job.Spec.HyperParameterTuningJobConfig.ResourceLimits.MaxNumberOfTrainingJobs + 5)
-			err = k8sClient.Update(context.Background(), job)
-			Expect(err).ToNot(HaveOccurred())
-
+			shouldHaveFinalizer = true
 		})
 
-		It("should set status to failed", func() {
-			// Not failed
-			description.HyperParameterTuningJobStatus = sagemaker.HyperParameterTuningJobStatusInProgress
-
-			// Setup mock responses.
-			sageMakerClient := builder.
-				WithRequestList(&receivedRequests).
-				AddDescribeHyperParameterTuningJobResponse(description).
-				Build()
-
-			// Instantiate controller and reconciliation request.
-			controller := createHpoReconcilerForSageMakerClient(k8sClient, sageMakerClient, "1s")
-			request := CreateReconciliationRequest(job.ObjectMeta.Name, job.ObjectMeta.Namespace)
-
-			// Run test and verify expectations.
-			controller.Reconcile(request)
-
-			// Verify status is failed.
-			err = k8sClient.Get(context.Background(), types.NamespacedName{
-				Namespace: job.ObjectMeta.Namespace,
-				Name:      job.ObjectMeta.Name,
-			}, job)
-			Expect(job.Status.HyperParameterTuningJobStatus).To(Equal(string(sagemaker.HyperParameterTuningJobStatusFailed)))
+		Context("HyperParameterTuningJob has status 'InProgress'", func() {
+			BeforeEach(func() {
+				expectedStatus = sagemaker.HyperParameterTuningJobStatusInProgress
+				mockSageMakerClientBuilder.
+					AddDescribeHyperParameterTuningJobResponse(CreateDescribeOutputWithOnlyStatus(expectedStatus))
+			})
+
+			When("!HasDeletionTimestamp", func() {
+				It("Requeues after interval", func() {
+					ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+				})
+
+				It("Updates status", func() {
+					ExpectStatusToBe(tuningJob, string(expectedStatus))
+				})
+
+				It("Attempts to spawn missing Training Jobs", func() {
+					ExpectSpawnMissingTrainingJobs(*jobSpawner, 1)
+				})
+
+				Context("Does not have a finalizer", func() {
+					BeforeEach(func() {
+						shouldHaveFinalizer = false
+					})
+
+					It("Adds a finalizer", func() {
+						ExpectToHaveFinalizer(tuningJob, controllers.SageMakerResourceFinalizerName)
+					})
+				})
+			})
+
+			When("HasDeletionTimestamp", func() {
+				BeforeEach(func() {
+					shouldHaveDeletionTimestamp = true
+					expectedStatus = sagemaker.HyperParameterTuningJobStatusStopping
+					mockSageMakerClientBuilder.
+						AddStopHyperParameterTuningJ
```

Comment:
We should have a test for when the delete of spawned training job fails. The controller needs to retry, otherwise it will leave a bunch of jobs around that have to be deleted manually (i.e. a special finalizer removed, then deleted).